Step3 - 질문 삭제하기 리팩터링

KangJunJun opened the pull request:
안녕하세요 영훈님! 
늦은 시간에 PR 드려 죄송합니다

작업하는데 생각보다 많이 딜레이가 된것 같습니다.

잘부탁드립니다

KangJunJun committed:
refactor: Step2 피드백 관련사항 수정

- deleted_test 에 보완 사항 있음

KangJunJun committed:
docs: Step3 기능목록 추가

KangJunJun committed:
feat: 질문과 답변 삭제시 실제 제거가 아닌 상태 변경으로 수정기능 구현

- 각 도메인에 delete 메서드 추가
- Test 코드 수정

KangJunJun committed:
feat: 질문과 답변 삭제 이력에 대한 정보를 DeleteHistory 에 저장

KangJunJun committed:
test: 테스트 픽스쳐 추가

KangJunJun committed:
refactor: List 를 랩핑하도록 클래스 추가

KangJunJun committed:
refactor: delete 로직 service에 적용

KangJunJun committed:
feat: 로그인 사용자인 경우만 삭제 가능하도록 구현

KangJunJun committed:
feat: 다른 사람이 쓴 답변이 있으면 삭제하지 못하도록 구현

chae-yh commented on README.md:
```diff
@@ -97,4 +97,18 @@ alter table question
+
+## STEP 3 - 질문 삭제하기 리팩터링
+
+### 기능 요구 사항
```
기능 요구 사항 작성 좋습니다 👍

chae-yh commented on src/main/java/qna/CannotDeleteException.java:
```diff
@@ -1,6 +1,6 @@
 package qna;
 
-public class CannotDeleteException extends Exception {
+public class CannotDeleteException extends RuntimeException {
```
여기를 unchecked exception으로 바꾸신 이유를 여쭤보고 싶습니다!

chae-yh commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -12,10 +13,10 @@ public class Answer extends BaseEntity {
     private User writer;
-    @ManyToOne(fetch = FetchType.LAZY)
-    @JoinColumn(name="question_id", foreignKey = @ForeignKey(name = "fk_answer_to_question"))
+    @ManyToOne(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
```
Answer가 만들어지는 시점에는 거의 Question이 만들어져 있을 것 같아요..! 혹시 CascadeType을 PERSIST로 주신 이유가 있을까요...?

chae-yh commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -41,16 +42,22 @@ public Answer(Long id, User writer, Question question, String contents) {
 
-    public boolean isOwner(User writer) {
-        return this.writer.equals(writer);
+    private void setQuestion(Question question) {
```
편의 관계 메서드 👍

chae-yh commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -72,12 +79,30 @@ public void setDeleted(boolean deleted) {
         return question;
     }
+
+    public DeleteHistory delete() {
```
이 메서드는 테스트에서만 사용하는 것 같아요...! 비지니스 로직에 사용하지 않는 메서드는 삭제하면 어떨까요..?

chae-yh commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -72,12 +79,30 @@ public void setDeleted(boolean deleted) {
+        return new DeleteHistory(ContentType.ANSWER, id, loginUser);
+    }
+
+    private void validateUser(User loginUser) {
```
잘 위치시켜 주셨습니다 👍

chae-yh commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -72,12 +79,30 @@ public void setDeleted(boolean deleted) {
                 "id=" + id +
                 ", writer=" + writer +
-                ", questionId=" + question.getId() +
+                ", question=" + question +
```
양방향 연관관계를 가지면 양방향 해당 형태면 양방향참조를 할 수도 있을 것 같아요. 기존처럼 id만 노출하면 어떨까요...?

chae-yh commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,40 @@
+import javax.transaction.Transactional;
+
+@Embeddable
+@Transactional
```
혹시 여기에 `@Transactional 는 어떤 사유로 붙이셨을까요..?

chae-yh commented on src/main/java/qna/domain/DeleteHistories.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class DeleteHistories {
```
일급 컬렉션 👍

chae-yh commented on src/main/java/qna/domain/Question.java:
```diff
@@ -17,8 +17,8 @@ public class Question extends BaseEntity {
     private User writer;
-    @OneToMany(mappedBy = "question", fetch = FetchType.LAZY, cascade = CascadeType.REMOVE)
-    private final List<Answer> answers = new ArrayList<>();
+    @Embedded
```
적절한 어노테이션 👍

chae-yh commented on src/main/java/qna/domain/Question.java:
```diff
@@ -45,8 +46,9 @@ public boolean isOwner(User writer) {
     public void addAnswer(Answer answer) {
-        answers.add(answer);
-        answer.toQuestion(this);
+        if (!answers.contains(answer)) {
```
아.. 이 부분에서 validation 체크를 해주었군요...! answers에서도 해줄 수 있을 것 같아요 ㅎㅎ

chae-yh commented on src/main/java/qna/domain/Question.java:
```diff
@@ -45,8 +46,9 @@ public boolean isOwner(User writer) {
 
     public void addAnswer(Answer answer) {
-        answers.add(answer);
-        answer.toQuestion(this);
```
answer에 대해서 Question을 set해주는 해당 메서드는 있어야 하지 않을까요..?

chae-yh commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -37,24 +34,6 @@ public Question findQuestionById(Long id) {
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
-        deleteHistoryService.saveAll(deleteHistories);
+        deleteHistoryService.saveAll(question.delete(loginUser).getList());
```
여기서 getList라는 getter를 이용해서 데이터를 가져오지 말고, 바로 처리해줄 수도 있을 것 같아요...!
service layer를 줄인 것은 좋습니다 👍

chae-yh commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,40 @@
+
+@Embeddable
+@Transactional
+public class Answers {
```
1급 컬렉션 👍

chae-yh commented on src/test/java/qna/domain/AnswersTest.java:
```diff
@@ -0,0 +1,40 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+public class AnswersTest {
```
도메인 추가에 따른 테스트 추가 👍

chae-yh commented on src/test/java/qna/domain/TestFixture.java:
```diff
@@ -0,0 +1,28 @@
+package qna.domain;
+
+public class TestFixture {
```
Fixture로 뽑아낸 것도 좋네요 👍

chae-yh commented on src/test/java/qna/repository/AnswerRepositoryTest.java:
```diff
@@ -76,7 +76,27 @@ void answer_to_question_lazy() {
+
+    @Test
+    @DisplayName("다른사람이 작성한 경우 예외발생")
+    void answer_delete_user_valid() {
```
꼼꼼한 테스트 좋습니다 👍

chae-yh commented on src/test/java/qna/repository/QuestionRepositoryTest.java:
```diff
@@ -61,20 +65,41 @@ void setWriter() {
 
     @DisplayName("질문삭제시 답변도 삭제 검증")
     @Test
     void deleted_test() {
```
꼼꼼한 테스트 좋네요 👍 영속성 컨텍스트 내에서의 테스트만이 아니라 DB까지 포함한 테스트여서 더 좋은 것 같아요 ㅎㅎ

chae-yh commented on src/test/java/qna/service/QnaServiceTest.java:
```diff
@@ -48,7 +48,7 @@ public void setUp() throws Exception {
     public void delete_성공() throws Exception {
         when(questionRepository.findByIdAndDeletedFalse(question.getId())).thenReturn(Optional.of(question));
-        when(answerRepository.findByQuestionIdAndDeletedFalse(question.getId())).thenReturn(Arrays.asList(answer));
+        //when(answerRepository.findByQuestionIdAndDeletedFalse(question.getId())).thenReturn(Arrays.asList(answer));
```
혹시 주석은 왜 추가된 건지 알 수 있을까요...?

chae-yh commented on src/test/java/qna/domain/AnswersTest.java:
```diff
@@ -0,0 +1,40 @@
+import org.junit.jupiter.api.Test;
+
+public class AnswersTest {
+    public static Answer answer;
```
혹시 여기에 static으로 주신 이유가 있을까요...?

chae-yh reviewed (changes requested):
준호님 안녕하세요. 벌써 jpa 마지막 미션이라니 아쉽네요..! 그리고 코드도 전체적으로 잘 작성해주셨습니다 👍 다만 조금 더 고민해주시면 좋을만한 내용이 있어 코멘트 드리니 확인 부탁드립니다.

추가로, 늦게 주시는 것도 전혀 괜찮습니다 ㅎㅎ 그리고, 작업 시 딜레이는 크게 문제되지 않을 것 같아요...! 천천히 하시더라도 많이 배우는 것이 더 좋은 방향인 것 같습니다 ㅎㅎ

내용 확인 부탁드립니다 감사합니다 : )

KangJunJun committed:
refactor: 피드백 반영

- 사용하지 않는 메서드 제거
- Unchecked Exception 를 다시 Checked Exception로 변경
- 의미없는 cascade 제거
- AnswersTest: BeforeAll를 BeforeEach로 바꾸고 static 제거

KangJunJun committed:
refactor: addAnswer의 valid 책임을 Answers로 이동

KangJunJun committed:
refactor: DeleteHistories add 로직 변경

- getter 메서드 제거
- front add 를 위해 addQueue 메서드로 변경

KangJunJun committed:
refactor: toString 양방향 참조 제거

KangJunJun committed:
test: 사용하지 않는 mocking 및 throws Exception 제거

chae-yh commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -35,26 +31,8 @@ public Question findQuestionById(Long id) {
 
     @Transactional
-    public void deleteQuestion(User loginUser, Long questionId) throws CannotDeleteException {
+    public void deleteQuestion(User loginUser, Long questionId) throws Exception {
```
여기 Exception을 기존 CannotDeleteException  -> 변경 Exception 하신 이유가 있으실까요...?

chae-yh commented on src/main/java/qna/domain/DeleteHistories.java:
```diff
@@ -0,0 +1,23 @@
+    public List<DeleteHistory> addQueue(DeleteHistory deleteHistory) {
+        Collections.reverse(deleteHistories);
+        deleteHistories.add(deleteHistory);
+        Collections.reverse(deleteHistories);
```
Collections.reverse 를 두 번 해주는 이유가 있을까요...?

chae-yh reviewed (approved):
준호님 안녕하세요! 전체적으로 피드백 잘 반영해주셨습니다...! 궁금한 점이 몇가지 있어서 코멘트 드렸으나, 해당 step의 목적과 직접적으로 관련은 없는 것으로 보여 여기서 merge 하려 합니다. 시간 되시면 내용 확인해주시면 좋을 것 같아요.

그리고 여기까지가 제가 준호님께 드리는 마지막 리뷰가 될 것 같습니다. 나름대로는 열심히 리뷰 드리려고 노력했는데, 준호님께 도움이 조금이라도 되었다면 좋겠네요 ㅎㅎ

남음 미션에서도 많은 것을 얻어가길 바랍니다..! 고생많으셨습니다. 감사합니다 : )
